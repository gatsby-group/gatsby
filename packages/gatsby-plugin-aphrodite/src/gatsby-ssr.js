import React from 'react'
import { renderToString } from 'react-dom/server'
import { StyleSheetServer } from 'aphrodite'

exports.replaceRenderer = ({
  bodyComponent,
  replaceBodyHTMLString,
  setHeadComponents,
}) => {
  const { html, css } = StyleSheetServer
    .renderStatic(() => renderToString(bodyComponent))

  replaceBodyHTMLString(html)

  setHeadComponents([
    <style
      id="aphrodite-styles"
      key="aphrodite-styles"
      dangerouslySetInnerHTML={{ __html: css.content }}
    />,
    <script
      id="aphrodite-ids"
      key="aphrodite-ids"
      dangerouslySetInnerHTML={{
        __html: `
        // <![CDATA[
        window._aphrodite = ${JSON.stringify(css.renderedClassNames)}
        // ]]>
        `,
      }}
    />,
  ])
}
