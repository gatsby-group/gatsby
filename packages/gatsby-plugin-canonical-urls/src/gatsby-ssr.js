import React from "react"
import url from "url"

exports.onRenderBody = (
  { setHeadComponents, pathname = `/` },
  pluginOptions
) => {
  if (pluginOptions && pluginOptions.siteUrl) {
    const parsedUrl = url.parse(pluginOptions.siteUrl)
    let myUrl = `${pluginOptions.siteUrl}${pathname}`
    console.log(`What are plugin options ${JSON.stringify(pluginOptions)}`)
    if(pluginOptions.requireTrailingSlash || false) {
      console.log(`Requiring trailing slash on url ${myUrl}`)
      myUrl = myUrl.replace(/\/$|$/, `/`)
      console.log(`Finished: ${myUrl}`)
    }
    setHeadComponents([
      <link
        rel="canonical"
        key={myUrl}
        href={myUrl}
        data-baseprotocol={parsedUrl.protocol}
        data-basehost={parsedUrl.host}
      />,
    ])
  }
}
