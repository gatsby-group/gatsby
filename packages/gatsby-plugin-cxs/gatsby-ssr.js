"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _react = _interopRequireDefault(require("react"));

var _cxs = _interopRequireDefault(require("cxs"));

var _jsxFileName = "/Users/ctxhou/Program/Github/gatsby/packages/gatsby-plugin-cxs/src/gatsby-ssr.js";

exports.onRenderBody = function (_ref) {
  var setHeadComponents = _ref.setHeadComponents;

  var css = _cxs.default.css();

  setHeadComponents([_react.default.createElement("style", {
    id: "cxs-ids",
    key: "cxs-ids",
    dangerouslySetInnerHTML: {
      __html: css
    },
    __source: {
      fileName: _jsxFileName,
      lineNumber: 7
    },
    __self: this
  })]);
};