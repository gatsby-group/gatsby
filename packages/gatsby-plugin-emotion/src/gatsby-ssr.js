const React = require(`react`)
const { renderToString } = require(`react-dom/server`)
const { extractCritical } = require(`emotion-server`)

const { wrapElement } = require(`./wrap-element`)

exports.replaceRenderer = ({
  bodyComponent,
  replaceBodyHTMLString,
  setHeadComponents,
}) => {
  const { html, ids, css } = extractCritical(
    renderToString(wrapElement(bodyComponent))
  )

  setHeadComponents([
    // eslint-disable-next-line react/jsx-key
    <style
      data-emotion-css={ids.join(` `)}
      dangerouslySetInnerHTML={{ __html: css }}
    />,
  ])

  replaceBodyHTMLString(html)
}
