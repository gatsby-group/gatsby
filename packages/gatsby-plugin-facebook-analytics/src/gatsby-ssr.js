import React from "react"
import { stripIndent } from "common-tags"

exports.onRenderBody = ({ setPostBodyComponents }, pluginOptions) => {
  if (
    process.env.NODE_ENV === `production` ||
    pluginOptions.includeInDevelopment
  ) {
    setPostBodyComponents([
      <script
        key="plugin-google-tagmanager"
        dangerouslySetInnerHTML={{
          __html: stripIndent`
          window.fbAsyncInit = function() {
            FB.init({
              appId      : ${pluginOptions.appId},
              xfbml      : true,
              version    : 'v2.12'
            });

            FB.AppEvents.logPageView();

          };

          (function(d, s, id){
             var js, fjs = d.getElementsByTagName(s)[0];
             if (d.getElementById(id)) {return;}
             js = d.createElement(s); js.id = id;
             js.src = "https://connect.facebook.net/en_US/sdk.js";
             fjs.parentNode.insertBefore(js, fjs);
           }(document, 'script', 'facebook-jssdk'));`,
        }}
      />,
    ])
  }
}
