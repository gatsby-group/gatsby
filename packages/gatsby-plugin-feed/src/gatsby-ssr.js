import React from "react"
import { withPathPrefix, withAssetPrefix } from "gatsby"
import { defaultOptions } from "./internals"

// TODO: remove for v3
if (!withAssetPrefix) {
  withAssetPrefix = withPathPrefix
}

exports.onRenderBody = ({ setHeadComponents }, pluginOptions) => {
  const { feeds } = {
    ...defaultOptions,
    ...pluginOptions,
  }

  const links = feeds.map(({ output, title }, i) => {
    if (output.charAt(0) !== `/`) {
      output = `/` + output
    }

    return (
      <link
        key={`gatsby-plugin-feed-${i}`}
        rel="alternate"
        type="application/rss+xml"
        title={title}
        href={withAssetPrefix(output)}
      />
    )
  })

  setHeadComponents(links)
}
