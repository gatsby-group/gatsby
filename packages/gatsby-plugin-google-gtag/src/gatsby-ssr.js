// @ts-check

import React from "react"
import { Minimatch } from "minimatch"

/**
 * @type {import('gatsby').GatsbySSR["onRenderBody"]}
 */
exports.onRenderBody = (
  { setHeadComponents, setPostBodyComponents },
  { trackingIds, gtagConfig, pluginConfig }
) => {
  if (
    process.env.NODE_ENV !== `production` &&
    process.env.NODE_ENV !== `test`
  ) {
    return null
  }

  // Lighthouse recommends pre-connecting to google tag manager
  setHeadComponents([
    <link
      rel="preconnect"
      key="preconnect-google-gtag"
      href={pluginConfig.origin}
    />,
    <link
      rel="dns-prefetch"
      key="dns-prefetch-google-gtag"
      href={pluginConfig.origin}
    />,
  ])

  // Prevent duplicate or excluded pageview events being emitted on initial load of page by the `config` command
  // https://developers.google.com/analytics/devguides/collection/gtagjs/#disable_pageview_tracking

  gtagConfig.send_page_view = false

  const firstTrackingId = trackingIds[0]

  const excludeGtagPaths = []
  if (pluginConfig.exclude.length > 0) {
    pluginConfig.exclude.map(exclude => {
      const mm = new Minimatch(exclude)
      excludeGtagPaths.push(mm.makeRe())
    })
  }

  const setComponents = pluginConfig.head
    ? setHeadComponents
    : setPostBodyComponents

  const renderHtml = () => `
      ${
        excludeGtagPaths.length
          ? `window.excludeGtagPaths=[${excludeGtagPaths.join(`,`)}];`
          : ``
      }
      ${
        gtagConfig.anonymize_ip === true
          ? `function gaOptout(){document.cookie=disableStr+'=true; expires=Thu, 31 Dec 2099 23:59:59 UTC;path=/',window[disableStr]=!0}var gaProperty='${firstTrackingId}',disableStr='ga-disable-'+gaProperty;document.cookie.indexOf(disableStr+'=true')>-1&&(window[disableStr]=!0);`
          : ``
      }
      if(${
        pluginConfig.respectDNT === true
          ? `!(navigator.doNotTrack == "1" || window.doNotTrack == "1")`
          : `true`
      }) {
        window.dataLayer = window.dataLayer || [];
        function gtag(){dataLayer.push(arguments);}
        gtag('js', new Date());

        ${trackingIds
          .map(
            trackingId =>
              `gtag('config', '${trackingId}', ${JSON.stringify(gtagConfig)});`
          )
          .join(``)}
      }
      `

  return setComponents([
    <script
      key={`gatsby-plugin-google-gtag`}
      async
      src={`${pluginConfig.origin}/gtag/js?id=${firstTrackingId}`}
    />,
    <script
      key={`gatsby-plugin-google-gtag-config`}
      dangerouslySetInnerHTML={{ __html: renderHtml() }}
    />,
  ])
}
