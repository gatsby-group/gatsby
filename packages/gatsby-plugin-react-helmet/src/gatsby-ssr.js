import { Helmet } from "react-helmet"

export const onRenderBody = ({
  setHeadComponents,
  setHtmlAttributes,
  setBodyAttributes,
}) => {
  const helmet = Helmet.renderStatic()
  // These action functions were added partway through the Gatsby 1.x cycle.
  if (setHtmlAttributes) {
    setHtmlAttributes(helmet.htmlAttributes.toComponent())
  }
  if (setBodyAttributes) {
    setBodyAttributes(helmet.bodyAttributes.toComponent())
  }
  setHeadComponents([
    helmet.title.toComponent(),
    helmet.link.toComponent(),
    helmet.meta.toComponent(),
    helmet.noscript.toComponent(),
    helmet.script.toComponent(),
    helmet.style.toComponent(),
    helmet.base.toComponent(),
  ])
}

export const onPreRenderHTML = ({
  getHeadComponents,
  replaceHeadComponents,
}) => {
  const headComponents = getHeadComponents()

  const groupedByType = headComponents.reduce((hash, obj) => {
    if (obj.type === undefined) return hash
    return Object.assign(hash, {
      [obj.type]: (hash[obj.type] || []).concat(obj),
    })
  }, {})

  replaceHeadComponents([
    ...groupedByType.title,
    ...groupedByType.base,
    ...groupedByType.meta,
    ...groupedByType.style,
    ...groupedByType.link,
    ...groupedByType.script,
    ...groupedByType.noscript,
  ])
}
