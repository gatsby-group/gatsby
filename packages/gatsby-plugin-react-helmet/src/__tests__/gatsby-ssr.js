import { onRenderBody } from "../gatsby-ssr"

jest.mock(`react-helmet`)

describe(`gatsby-plugin-react-helmet`, () => {
  describe(`onRenderBody`, () => {
    it(`sets head components`, () => {
      const setHeadComponents = jest.fn()
      onRenderBody({ setHeadComponents })
      expect(setHeadComponents.mock.calls).toMatchSnapshot()
    })

    it(`sets html attributes`, () => {
      const setHeadComponents = jest.fn()
      const setHtmlAttributes = jest.fn()
      onRenderBody({ setHeadComponents, setHtmlAttributes })
      expect(setHtmlAttributes.mock.calls).toMatchSnapshot()
    })

    it(`sets body attributes`, () => {
      const setHeadComponents = jest.fn()
      const setBodyAttributes = jest.fn()
      onRenderBody({ setHeadComponents, setBodyAttributes })
      expect(setBodyAttributes.mock.calls).toMatchSnapshot()
    })
  })
})
