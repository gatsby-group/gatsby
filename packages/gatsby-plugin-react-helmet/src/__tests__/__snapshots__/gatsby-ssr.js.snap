// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gatsby-plugin-react-helmet onRenderBody sets body attributes 1`] = `
Array [
  Array [
    "body-attributes-component",
  ],
]
`;

exports[`gatsby-plugin-react-helmet onRenderBody sets head components 1`] = `
Array [
  Array [
    Array [
      "title-component",
      "link-component",
      "meta-component",
      "noscript-component",
      "script-component",
      "style-component",
    ],
  ],
]
`;

exports[`gatsby-plugin-react-helmet onRenderBody sets html attributes 1`] = `
Array [
  Array [
    "html-attributes-component",
  ],
]
`;
